### PR TITLE
Add grow-trait support for crops | Farmer job

### LINKED ISSUE #44: Job Block Meta
Currently if something like wheat has a reward for it, the player can keep placing it down and destroying it for a reward without having to wait for it to be fully grown. A way of preventing this and checking if it is fully grown is needed.

### PR DESCRIPTION
**Pull request by MarkL4YG**
Hey there,

It's me again

Description
---
This PR is supposed to resolve #44 by adding support for a numeric block trait that will be read as indicating the current growth (state?) of a plant. For example age the ``age`` ``BlockTrait<Integer>`` of vanilla crops like wheat and cocoa.  The reward is calculated by taking a percentage of the full reward based on the growth state of the block.  
Also included is a change in the exploit prevention to allow growing blocks to be placed by players and generate income when broken.  
There's also a change when reading the ``JobSet``s upon a tracked event: If multiple ``JobSet``s provide rewards for the same action and id TE uses the one granting more exp-gain instead of just the last one that's been read.

I also tried taking a look at if it's possible to limit the reward to a specific tool being used but as far as I saw there's no reliable way to get the tool used for break-events so I left that open for future development.

Related issues
---
* #44 

Tested on
---
SpongeForge: stable-5
SpongeVanilla: stable-5

### CODE REVIEW

**Comment by MarkL4YG** [comment]
Oh and if someone comes up with a better way for updating the fork - feel free to tell me ^^

File: src/main/java/com/erigitic/jobs/jobsets/FarmerJobSet.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Override
+    public void populateNode(ConfigurationNode node) {
+        ConfigurationNode myNode = node.getNode(SETNAME);
+
+        for (String[] a : REWARDS) {
+            ConfigurationNode n = myNode.getNode(a[0], a[1]);
+            n.getNode("exp").setValue(a[2]);
+            n.getNode("money").setValue(a[3]);
+            n.getNode("growTrait").setValue(a[4]);
```
Comment (Erigitic) [inline src/main/java/com/erigitic/jobs/jobsets/FarmerJobSet.java]: Change to `growthTrait` to match the variable name in TEActionReward

File: src/main/java/com/erigitic/jobs/TEJobManager.java [hunk 2]
```diff
@@ -677,41 +682,72 @@ public void onPlayerBlockBreak(ChangeBlockEvent.Break event) {
...
+                List<String> sets = optPlayerJob.get().getSets();
 
-                        reward = optSet.get().getRewardFor("break", blockName);
+                for (String s : sets) {
+                    Optional<TEJobSet> optSet = getJobSet(s);
+                    if (!optSet.isPresent()) {
+                        logger.warn("Job " + playerJob + " has the nonexistent set \"" + s + "\"");
+                        continue;
                     }
+                    Optional<TEActionReward> cr = optSet.get().getRewardFor("break", blockName);
```
Comment (Erigitic) [inline src/main/java/com/erigitic/jobs/TEJobManager.java]: Change from `cr` to something more descriptive such as `currentReward` or to whatever you think fits best.

File: src/main/java/com/erigitic/jobs/jobsets/FarmerJobSet.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            {"break", "minecraft:cocoa", "30", "60.00", "age"},
+
+            {"break", "minecraft:cactus", "10", "15.00", null},
+            {"break", "minecraft:reeds", "5", "7.50", null},
+
+            //These plants aren't hard to grow or harvest thus the low gains
+            {"break", "minecraft:red_flower", "1", "0.10", null},
+            {"break", "minecraft:yellow_flower", "1", "0.10", null},
+            {"break", "minecraft:vine", "0", "0.05", null},
+            {"break", "minecraft:waterlily", "1", "0.1", null},
```
Comment (Erigitic) [inline src/main/java/com/erigitic/jobs/jobsets/FarmerJobSet.java]: Change money reward from `0.1` to `0.10` to match the significant figures from the rest

File: src/main/java/com/erigitic/jobs/jobsets/FarmerJobSet.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+            {"break", "minecraft:red_mushroom", "8", "5", null},
```
Comment (Erigitic) [inline src/main/java/com/erigitic/jobs/jobsets/FarmerJobSet.java]: Change from `5` to `5.00`

File: src/main/java/com/erigitic/jobs/jobsets/FarmerJobSet.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            {"break", "minecraft:brown_mushroom", "8", "5", null}
```
Comment (Erigitic) [inline src/main/java/com/erigitic/jobs/jobsets/FarmerJobSet.java]: Change from `5` to `5.00`

File: src/main/java/com/erigitic/jobs/TEJobManager.java [hunk 6]
```diff
@@ -677,41 +682,72 @@ public void onPlayerBlockBreak(ChangeBlockEvent.Break event) {
...
-                        TEAccount playerAccount = (TEAccount) accountManager.getOrCreateAccount(player.getUniqueId()).get();
-
-                        if (notify) {
-                            notifyPlayer(player, payAmount);
+                if (reward.isPresent()) {
+                    int expAmount = reward.get().getExpReward();
+                    BigDecimal payAmount = reward.get().getMoneyReward();
+                    Optional<String> growTrait = reward.get().getGrowthTrait();
+                    // If there is a growth trait calculate a percentage to compensate only partly grown crops
+                    if (growTrait.isPresent()) {
```
Comment (Erigitic) [inline src/main/java/com/erigitic/jobs/TEJobManager.java]: Change variable name to `growthTrait`

**Comment by MarkL4YG** [comment]
I hope I got all

**Comment by MarkL4YG** [comment]
Nope I didn't

File: src/main/java/com/erigitic/jobs/TEJobManager.java [hunk 7]
```diff
@@ -808,14 +848,18 @@ public void onPlayerKillEntity(DestructEntityEvent.Death event) {
 
                     for (String s : sets) {
                         Optional<TEJobSet> optSet = getJobSet(s);
-
                         if (!optSet.isPresent()) {
-                            logger.warn("[TE] Job " + getPlayerJob(player) + " has nonexistent set \"" + s + "\"");
-
+                            logger.warn("Job " + playerJob + " has the nonexistent set \"" + s + "\"");
                             continue;
                         }
```
Comment (MarkL4YG) [inline src/main/java/com/erigitic/jobs/TEJobManager.java]: Exists - Outdated

File: src/main/java/com/erigitic/jobs/TEJobManager.java [hunk 8]
```diff
@@ -871,14 +915,18 @@ public void onPlayerFish(FishingEvent.Stop event) {
...
```
Comment (MarkL4YG) [inline src/main/java/com/erigitic/jobs/TEJobManager.java]: Exists - Outdated

**Comment by MarkL4YG** [comment]
You're welcome

**Pull request merged**